### PR TITLE
Add support for local language configuration

### LINKED ISSUE #1217: Local configuration folder
Hi everyone!
First, thank you for that amazing work!

#### Describe your feature request

I'm working on a big [codebase](https://github.com/vectordotdev/vector) that takes several minutes to build. So I've to disable features in order to make it run faster.
I can do that by passing arguments to `rust-analyzer` in the `languages.toml` config file.
The problem is that it applies for every helix instances.

I'd love to have a `.helix` folder in the project folder with a `languages.toml` file in it to disable features only for the current project.

### PR DESCRIPTION
**Pull request by kirawi**
Resolves #1217, allowing users to have configuration for local directories. It defaults to looking for every `.helix` folder until it finds a `.git` folder or is at the root, and merges all the configs in order.

### CODE REVIEW

File: helix-term/src/application.rs [hunk 1]
```diff
@@ -52,30 +52,57 @@ pub struct Application {
 impl Application {
     pub fn new(args: Args, mut config: Config) -> Result<Self, Error> {
         use helix_view::editor::Action;
-        let mut compositor = Compositor::new()?;
-        let size = compositor.size();
 
-        let conf_dir = helix_core::config_dir();
+        let config_dir = helix_core::config_dir();
+        let local_config_dir = helix_core::local_config_dir().unwrap();
```
Comment (kirawi) [inline helix-term/src/application.rs:57]: Note: Should instead unwrap in the `local_config_dir()` function.

Comment (archseer) [reply helix-term/src/application.rs:57]: There should be no unwraps, it's perfectly valid not to find a root. You could use cwd in that case or just do nothing.

File: helix-core/src/lib.rs [hunk 2]
```diff
@@ -92,6 +92,14 @@ pub fn config_dir() -> std::path::PathBuf {
     path
 }
 
+/// Searches for the local `.helix` directory by searching for the root `.git` directory,
+/// using the CWD if it can't find one.
+pub fn local_config_dir() -> std::path::PathBuf {
+    let root = find_root(None)
+        .unwrap_or_else(|| std::env::current_dir().expect("unable to determine current directory"));
+    root.join(".helix")
+}
```
Comment (sudormrfbin) [inline helix-core/src/lib.rs:101]: Shouldn't we check if the cwd contains a .helix directory first ? So in effect:

1. Explicitly check if cwd has `.helix`. If so use that. If not...
2. Search for the root dir. If found use `root/.helix`. If not...
3. Assume `cwd/.helix` (which would not exist).

I remember someone opening an issue about having a .git repo in their home dir and helix calculating the wrong root.

File: helix-core/src/lib.rs [hunk 3]
```diff
@@ -92,6 +92,19 @@ pub fn config_dir() -> std::path::PathBuf {
...
+/// to avoid an unwrap.
+pub fn local_config_dir() -> std::path::PathBuf {
+    let cwd = std::env::current_dir().expect("unable to determine current directory");
+    if cwd.join(".helix").is_dir() {
+        cwd.join(".helix")
+    } else {
+        let root = find_root(None).unwrap_or(cwd);
+        root.join(".helix")
+    }
+}
```
Comment (pickfire) [inline helix-core/src/lib.rs:106]: Wouldn't user to expect the local to overwrite root config if both exists? Shouldn't we read both and merge them with local being the last to be merged?

File: helix-term/src/application.rs [hunk 4]
```diff
@@ -52,77 +52,110 @@ pub struct Application {
...
-            }
-        } else {
-            theme_loader.default()
-        };
+        // These configuration directories can contain `config.toml` and `languages.toml`.
+        // `local_config_dir` is a `.helix` folder within the projec directory.
+        let config_dir = helix_core::config_dir();
+        let local_config_dir = helix_core::local_config_dir();
+
+        // Config override order: local -> global -> default.
```
Comment (pickfire) [inline helix-term/src/application.rs:61]:
```suggestion
        // Config override order: local (cwd) -> local (root) -> global -> default.
```
I think this is a better option.

File: helix-term/src/application.rs [hunk 5]
```diff
@@ -52,77 +52,110 @@ pub struct Application {
...
+                    Err(err) => return Err(Error::new(err)),
+                };
+                let global_config = match std::fs::read(config_dir.join("languages.toml")) {
+                    Ok(config) => toml::from_slice(&config)
+                        .expect("failed to read the global `languages.toml`"),
+                    Err(err) if err.kind() == std::io::ErrorKind::NotFound => {
+                        toml::Value::Table(toml::value::Table::default())
+                    }
+                    Err(err) => return Err(Error::new(err)),
+                };
```
Comment (pickfire) [inline helix-term/src/application.rs:83]: With the suggestion I mentioned, there should be some kind of code reuse for this.

File: helix-term/src/main.rs [hunk 6]
```diff
@@ -85,12 +85,13 @@ FLAGS:
...
-    if !conf_dir.exists() {
-        std::fs::create_dir_all(&conf_dir).ok();
+    let config_dir = helix_core::config_dir();
+    if !config_dir.exists() {
+        std::fs::create_dir_all(&config_dir).ok();
     }
 
-    let config = match std::fs::read_to_string(conf_dir.join("config.toml")) {
+    // `config.toml`
+    let config = match std::fs::read_to_string(config_dir.join("config.toml")) {
```
Comment (archseer) [inline helix-term/src/main.rs:94]: Why not move this into application.rs now?

File: helix-term/src/application.rs [hunk 7]
```diff
@@ -49,80 +49,113 @@ pub struct Application {
...
+        // Read and parse the `languages.toml` files as TOML objects.
+        let default_lang_config: toml::Value =
+            toml::from_slice(include_bytes!("../../languages.toml"))
+                .expect("failed to read the default `languages.toml`");
+        let lang_config = {
+            let local_cwd_config = load_lang_config(cwd.join(".helix/languages.toml"))?;
+            let local_root_config = load_lang_config(local_config_dir.join("languages.toml"))?;
+            let global_config = load_lang_config(config_dir.join("languages.toml"))?;
+
+            merge_toml_values(
```
Comment (archseer) [inline helix-term/src/application.rs:82]: Use an array and do a .fold

File: helix-term/src/main.rs (hunk 6 above)
Comment (kirawi) [reply helix-term/src/main.rs:94]: I wasn't sure if it would interfere with the test harness you're working on.

File: helix-term/src/application.rs [hunk 8]
```diff
@@ -49,33 +49,94 @@ pub struct Application {
...
+            let local_cwd_config = load_lang_config(cwd.join(".helix/languages.toml"))?;
+            let local_root_config = load_lang_config(local_config_dir.join("languages.toml"))?;
+            let global_config = load_lang_config(config_dir.join("languages.toml"))?;
+            [
+                local_root_config,
+                global_config,
+                default_lang_config.clone(),
+            ]
+            .into_iter()
+            .fold(local_cwd_config, merge_toml_values)
```
Comment (kirawi) [inline helix-term/src/application.rs:106]: Ah I just realized I messed up the order here.

**Comment by kirawi** [comment]
Tested and seems to work.

File: helix-core/src/lib.rs [hunk 9]
```diff
@@ -93,6 +93,15 @@ pub fn config_dir() -> std::path::PathBuf {
     path
 }
 
+/// First looks for the local `.helix` directory in CWD. If it can't find it, then
+/// it searches for it within the root `.git` directory. Otherwise, it uses the CWD
+/// to avoid an unwrap.
```
Comment (sudormrfbin) [inline helix-core/src/lib.rs:98]: This comment seems to be outdated.

File: helix-term/src/application.rs [hunk 10]
```diff
@@ -49,33 +49,94 @@ pub struct Application {
...
 
+        // Config override order: local (cwd) -> local (root) -> global -> default.
+        // Read and parse the `languages.toml` files as TOML objects.
+        let default_lang_config: toml::Value =
+            toml::from_slice(include_bytes!("../../languages.toml"))
+                .expect("failed to read the default `languages.toml`");
+        let lang_config: toml::Value = {
+            let local_cwd_config = load_lang_config(cwd.join(".helix/languages.toml"))?;
+            let local_root_config = load_lang_config(local_config_dir.join("languages.toml"))?;
+            let global_config = load_lang_config(config_dir.join("languages.toml"))?;
```
Comment (sudormrfbin) [inline helix-term/src/application.rs:99]: Using strings here feels icky, I think top level constants would be better here for `languages.toml`, `../../languages.toml`, etc.

Comment (kirawi) [reply helix-term/src/application.rs]: Unfortunately a constant doesn't work for `include_bytes!` because it expects a string literal only.

File: helix-term/src/application.rs [hunk 11]
```diff
@@ -53,33 +53,97 @@ pub struct Application {
...
+            let local_cwd_config = load_lang_config(cwd.join(".helix").join(LANG_CONFIG_FILENAME))?;
+            let local_root_config = load_lang_config(local_config_dir.join(LANG_CONFIG_FILENAME))?;
+            let global_config = load_lang_config(config_dir.join(LANG_CONFIG_FILENAME))?;
+            [
+                local_root_config,
+                global_config,
+                default_lang_config.clone(),
+            ]
+            .into_iter()
+            .fold(local_cwd_config, |a, b| merge_toml_values(b, a))
```
Comment (pickfire) [inline helix-term/src/application.rs:113]: Why not iterate over paths rather than lang config? That way we don't have to `load_lang_config` so many times.

File: helix-term/src/application.rs [hunk 12]
```diff
@@ -101,42 +165,32 @@ impl Application {
...
+        // Handle CLI arguments.
         if args.load_tutor {
             let path = helix_core::runtime_dir().join("tutor.txt");
             editor.open(path, Action::VerticalSplit)?;
             // Unset path to prevent accidentally saving to the original tutor file.
             doc_mut!(editor).set_path(None)?;
         } else if !args.files.is_empty() {
-            let first = &args.files[0]; // we know it's not empty
+            // File paths passed as e.g. `hx foo.rs bar.rs`
+            // SAFETY: The file count is already known to be non-zero.
```
Comment (pickfire) [inline helix-term/src/application.rs:183]: I thought `SAFETY` is used for `unsafe`? Using it here feels a bit weird.

**Comment by kirawi** [comment]
I'm not sure how to handle errors as they don't say which file the error originated from. Ideas?

**Comment by kirawi** [comment]
Ready for review

**Comment by TotalKrill** [comment]
I have a rebased version in https://github.com/TotalKrill/helix/tree/local

File: helix-core/src/lib.rs [hunk 13]
```diff
@@ -56,32 +61,174 @@ pub fn find_root(root: Option<&str>, root_markers: &[String]) -> Option<std::pat
...
+pub fn lang_config_file() -> std::path::PathBuf {
+    config_dir().join("languages.toml")
+}
+
+pub fn log_file() -> std::path::PathBuf {
+    cache_dir().join("helix.log")
+}
+
+// right overrides left
+pub fn merge_toml_values(left: toml::Value, right: toml::Value) -> toml::Value {
```
Comment (archseer) [inline helix-core/src/lib.rs:146]: A bunch of these seem to duplicate code that was moved to helix-loader

File: helix-core/src/lib.rs [hunk 14]
```diff
@@ -45,7 +45,12 @@ pub fn find_first_non_whitespace_char(line: RopeSlice) -> Option<usize> {
 /// * Top-most folder containing a root marker if not git repository detected
 /// * Current working directory as fallback
 pub fn find_root(root: Option<&str>, root_markers: &[String]) -> Option<std::path::PathBuf> {
+    find_root_impl(root, root_markers).first().cloned()
+}
+
+fn find_root_impl(root: Option<&str>, root_markers: &[String]) -> Vec<std::path::PathBuf> {
     let current_dir = std::env::current_dir().expect("unable to determine current directory");
```
Comment (archseer) [inline helix-core/src/lib.rs:52]: Why the change here if you end up ignoring everything except the first element anyway?

File: helix-core/src/lib.rs (hunk 13 above)
Comment (kirawi) [reply helix-core/src/lib.rs:146]: Yeah I messed up the merge.

File: helix-core/src/lib.rs (hunk 14 above)
Comment (kirawi) [reply helix-core/src/lib.rs:52]: I needed `find_root_impl` to return a list of all directories that match `root` up to `.git` for `local_config_dirs()`

Comment (kirawi) [reply helix-core/src/lib.rs:52]: e.g.
```
.
└── project/
    ├── .git
    ├── .helix
    └── foo/
        └── .helix/
            └── bar/
                └── .helix
```

**Comment by poliorcetics** [comment]
While in no way a blocking concern, the more tools I use, the more `.<tool>` there are in my git repos. Something like `.config/<tool1>`, `.config/<tool2>` would be very very nice.

I don't know if this is something that is easy to do, and even if it is, it may not be fast enough/maintainable enough. There may also be flaws I have not thought about for this

**Comment by kirawi** [comment]
I thought about that, but pretty much nobody else does it that way so in the end it would still be an extra folder, except there's now also nesting.

**Comment by poliorcetics** [comment]
I suppose it can always be added backward compatibly later, and I want this functionality too much to delay it more, I'll be fine with a .helix directly

**Comment by elferherrera** [comment]
This is really nice. What do you need for testing?

**Comment by archseer** [comment]
If it seems to work for users and there's no bugs then I'll give it another review this weekend

**Comment by elferherrera** [comment]
Ir works for me. I added a new `languages.toml` file in a .helix folder and it loaded that config. What I dont understand from the PR is if it will override all of one language settings or just sections of it. 

Also, could it be discussed to have a window to show LSP loaded parameters?

**Comment by the-mikedavis** [comment]
The merging behavior is on a per-key basis: if you want to turn LSP auto-format off for rust for example, you can have a `languages.toml` that says

```toml
[[language]]
name = "rust"
auto-format = false
```

And that will override that key but not change other keys like `comment-token`.

**Comment by kirawi** [comment]
I tested on:
```
a
├───.git
├───.helix
└───b
    ├───.helix
    └───c
        └───.helix
```
With:

`a/.helix/languages.toml`
```toml
[[language]]
name = "rust"
auto-format = false
```

`b/.helix/languages.toml`
```toml
[[language]]
name = "rust"
comment-token = "#"
```

`c/.helix/languages.toml`
```toml
[[language]]
name = "rust"
indent = { tab-width = 2, unit = "  " }
```

correctly becoming
```toml
[[language]]
auto-format = false
comment-token = "#"
file-types = ["rs", "rs"]
injection-regex = "rust"
name = "rust"
roots = ["Cargo.toml", "Cargo.lock", "Cargo.toml", "Cargo.lock"]
scope = "source.rust"
```

**Comment by elferherrera** [comment]
@kirawi how did you check the final result?

**Comment by kirawi** [comment]
I added `log::debug!("{}", config);` at https://github.com/helix-editor/helix/pull/1249/files#diff-75a83fbc5b7b51ab917504525e40d867b69ccfc3102d7f7cc1d90281382f5f86R24

**Comment by elferherrera** [comment]
I see. I thought there was an option I didn't know

**Review by archseer — APPROVED** [review]

File: helix-term/src/application.rs [hunk 15]
```diff
@@ -56,15 +56,33 @@ pub struct Application {
 }
 
 impl Application {
-    pub fn new(args: Args, config: Config) -> Result<Self, Error> {
+    pub fn new(args: Args) -> Result<Self, Error> {
```
Comment (dead10ck) [inline helix-term/src/application.rs:59]: I'm sorry I'm kind of late to the party here, this PR eluded me. This change actually presents a problem. I'm working on an integration testing branch, and this change makes it so that you can't pass in an anyway parsed Config object. This was very useful for testing, as it's super easy to just make the struct literal you want to describe the config option you want to test. Now it seems that the config can *only* come from a toml file. Could we please move this logic back into main?

**Comment by davidchisnall** [comment]
[ Sorry for commenting here, but you appear not to have set up the security tab for reporting potential  vulnerabilities ]

If I add a .helix folder in my git repo that sets the language server protocol command to `rm -rf ~` then won't a Helix user that opens a file inside a clone of my git repo end up deleting their entire home directory?  Most other systems that have similar behaviour have some mitigation such as allowing only commands from an allow list to run or requiring the user to inspect and authorise the config file every time that it changes.  I didn't see anything here.

**Comment by kirawi** [comment]
This isn't the appropriate place to make that comment. You should create a separate issue instead.

**Comment by davidchisnall** [comment]
If you do not configure the security settings on GitHub or provide a SECURITY.me, external people have no idea how to report suspected security issues. There was no coordinated disclosure process documented for this project, so I filed it on the place closest to where the problem was introduced. If you want security issues filed in a specific place then I suggest that you document this. Either way, I have reported it and it’s now up to you to choose how to handle it.

**Comment by pascalkuthe** [comment]
I dont think editors have a particular high thread model that would warrant a security policy like that. If you clone untrusted code and run any language server on it you are already vulnerable (rust analyzer runs cargo check which runs build scripts which can run arbitrary untrusted code, same for many other LS) so really any editor that runs language servers is gulernabke. So emacs-lsp mode, nvim and kakune can all be attacked if you open files from a compromised repo.

Either way just like any other problem it should be reported as an issue instead of a random comment on old PR that is likely easily forgotten.

**Comment by archseer** [comment]
I'm currently working on a patch to explicitly approve each new workspace.
